consensus/src/experimental/buffer_manager.rs

```diff
@@ -290,6 +365,117 @@ impl StateManager {
         }
     }
 
+    async fn process_successful_signing_resp(
+        &mut self,
+        sig: Ed25519Signature,
+        commit_ledger_info: LedgerInfo,
+    ) -> anyhow::Result<()> {
+        let mut current_cursor = self.signing_root.clone();
+        if current_cursor.is_some() {
+            // this is important because the responses might not come in order because
+            // retrying a failed signature and finishing execution will incur requests
+            // to signing phase
+            while current_cursor.is_some() {
+                // update signature
+                let buffer_item = take_elem(&current_cursor);
+                if !buffer_item.has_been_executed() {
+                    // a reset has happened
+                    // we do nothing except put it back
+                    set_elem(&current_cursor, buffer_item);
+                    break;
+                }
+
+                if buffer_item.get_commit_info() == commit_ledger_info.commit_info() {
+                    // it is possible that we already signed this buffer item (double check after the final integration)
+                    if matches!(buffer_item, BufferItem::Executed(_)) {
+                        // we have found the buffer item
+                        let (signed_buffer_item, commit_vote) =
+                            buffer_item.advance_to_signed(self.author, sig.clone(), &self.verifier);
+
+                        set_elem(&current_cursor, signed_buffer_item);
+
+                        // send out commit vote
+                        self.commit_msg_tx
+                            .broadcast(ConsensusMsg::CommitVoteMsg(Box::new(commit_vote)))
+                            .await;
+
+                        self.try_advance_signing_root();
+                    }
+                    break;
+                }
+                current_cursor = get_next(&current_cursor);
+            }
+        }
+        // otherwise, a reset happened, we do nothing
+        Ok(())
+    }
+
+    async fn process_signing_response(
+        &mut self,
+        signing_resp: SigningResponse,
+    ) -> anyhow::Result<()> {
+        let SigningResponse {
+            signature_result,
+            commit_ledger_info,
+        } = signing_resp;
+        if let Ok(sig) = signature_result {
+            self.process_successful_signing_resp(sig, commit_ledger_info)
+                .await
+        } else {
+            // try next signature if signing failure
```

Comment:
this is retrying the same block? why the comment says next?

Reply:
It is not the same buffer item. Signing root points to the first unprocessed item. But there might be Signed item scattered after the signing root. Below situation is possible:

[Signed] -> [Signed] -> [Executed] -> [Signed] -> [Executed] -> [Signed] -> [Executed] ...
And the signing root points to the third item. The failure could be related to the 4-th item.

This might happen because signing phase might not see the items in order (success execution response and failed signing response will both push an item to signing phase)